besu/src/main/java/org/hyperledger/besu/cli/util/BesuCommandCustomFactory.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.cli.util;
+
+import org.hyperledger.besu.services.BesuPluginContextImpl;
+
+import java.lang.reflect.Constructor;
+
+import picocli.CommandLine;
+
+/**
+ * Custom PicoCli IFactory to handle version provider construction with plugin versions. Based on
+ * same logic as PicoCLI DefaultFactory.
+ */
+public class BesuCommandCustomFactory implements CommandLine.IFactory {
+  private final BesuPluginContextImpl besuPluginContext;
+
+  public BesuCommandCustomFactory(final BesuPluginContextImpl besuPluginContext) {
+    this.besuPluginContext = besuPluginContext;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> T create(final Class<T> cls) throws Exception {
+    if (CommandLine.IVersionProvider.class.isAssignableFrom(cls)) {
+      // the besuPluginContext has registered required plugins by this time.
+      return (T) new VersionProvider(besuPluginContext.getPluginVersions());
```

Comment:
If `besuPluginContext.getPluginVersions()` can be null, should it be Optional.ofNullable?

Reply:
Empty array list.

Reply:
VersionProvider.java:31 does a null check.

Reply:
it was the defensive programmer in me :), but I will remove the null check as you are correct, the pluginlist in VersionProvider is not meant to be null.

Reply:
done.